<a href="https://colab.research.google.com/github/KhytsenkoIgor/python-data-analysis-tasks/blob/main/hw_10_4_apply%2C_groupby%2C_pivot_table.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df = pd.read_csv('drive/MyDrive/Модуль_10/data/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [10]:
profit_branch = df.groupby('Branch')['gross income'].sum()



In [14]:
profit_branch.sort_values(ascending=False)

,gross income
Branch,
C,5265.1765
A,5057.1605
B,5057.0320


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [17]:
profit_city = df.groupby('City')['gross income'].sum()

In [18]:
profit_city.sort_values(ascending=False)

,gross income
City,
Naypyitaw,5265.1765
Yangon,5057.1605
Mandalay,5057.0320


In [21]:
city_metrics = df.groupby('City').agg({
    'gross margin percentage': 'mean',
    'gross income': 'sum',
    'Rating': 'mean'})


In [22]:
city_metrics.sort_values('gross income', ascending=False)

,gross margin percentage,gross income,Rating
City,,,
Naypyitaw,4.761905,5265.1765,7.072866
Yangon,4.761905,5057.1605,7.027059
Mandalay,4.761905,5057.0320,6.818072


**Відповідь:** Naypyitaw виглядає найбільш перспективним містом для відкриття ще одного магазину. Воно має найбільший сумарний валовий дохід 5265,18, що вище ніж у Yangon та Mandalay. Також Naypyitaw має найвищий середній рейтинг покупців 7,07. Водночас відсоток валового прибутку в усіх трьох містах однаковий  4,76%, тому цей показник не дає Naypyitaw окремої переваги.За наявними даними Naypyitaw можна розглядати як кандидата для відкриття нового магазину, але для остаточного рішення потрібно було б врахувати й інші фактори, наприклад кількість покупців, обсяг продажів, витрати на відкриття магазину та рівень конкуренції.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [26]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')



In [27]:
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [28]:
invoices_by_category['Electronic accessories'].sort_values(ascending=False)

,Electronic accessories
Branch,
A,60
B,55
C,55


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**Крок 1. Гіпотези:**  

1.  Чи відрізняються способи оплати (Payment) між звичайними клієнтами та учасниками програми лояльності (Customer type)?
2.  Які категорії товарів дають найбільший обсяг продажів і валовий дохід?
3.  Чи змінюється задоволеність клієнта залежно від розміру покупки?






**Крок 2:**

In [38]:
    # Чи відрізняються способи оплати (Payment) між звичайними клієнтами та учасниками програми лояльності (Customer type)?

payment_by_customer = pd.pivot_table(df,index='Customer type',columns='Payment',values='Invoice ID',aggfunc='count')


In [39]:
payment_by_customer

Payment,Cash,Credit card,Ewallet
Customer type,,,
Member,168,172,161
Normal,176,139,184


In [33]:
    # Які категорії товарів дають найбільший обсяг продажів і валовий дохід?

product_analysis = df.groupby('Product line').agg({'Total':'sum','Quantity':'sum','gross income':'sum'})

product_analysis.sort_values('Total', ascending=False)

,Total,Quantity,gross income
Product line,,,
Food and beverages,56144.8440,952,2673.5640
Sports and travel,55122.8265,920,2624.8965
Electronic accessories,54337.5315,971,2587.5015
Fashion accessories,54305.8950,902,2585.9950
Home and lifestyle,53861.9130,911,2564.8530
Health and beauty,49193.7390,854,2342.5590


*Створюємо функцію **purchase_size**, яка визначає категорію розміру покупки за значенням Quantity: 1-3 товари -Small, 4 -7 -Medium, 8 -10 -Large.*

In [58]:
def purchase_size(quantity):     # Чи змінюється задоволеність клієнта залежно від розміру покупки?
    if quantity <= 3:
        return 'Small'
    elif quantity <= 7:
        return 'Medium'
    else:
        return 'Large'

*Створюємо нову колонку **Purchase size**, у яку за допомогою apply() записуємо результат роботи функції purchase_size для кожного значення колонки Quantity.*

In [59]:
df['Purchase size'] = df['Quantity'].apply(purchase_size)

In [61]:
df[['Quantity', 'Rating', 'Purchase size']]

,Quantity,Rating,Purchase size
0,7,9.1,Medium
1,5,9.6,Medium
2,7,7.4,Medium
3,8,8.4,Large
4,7,5.3,Medium
...,...,...,...
995,1,6.2,Small
996,10,4.4,Large
997,1,7.7,Small
998,1,4.1,Small


*Групуємо покупки за **Purchase size** та знаходимо середній **Rating** і середній **Total** для кожної групи*.

In [62]:
purchase_analysis = df.groupby('Purchase size').agg({'Rating':'mean','Total':'mean'})

purchase_analysis

,Rating,Total
Purchase size,,
Large,6.931081,537.25914
Medium,7.005839,317.32165
Small,6.968259,114.39879


**Крок 3:**



1.   **Чи відрізняються способи оплати між звичайними клієнтами та учасниками програми лояльності?**

*За результатами видно, що учасники програми лояльності найчастіше оплачують покупки кредитною карткою 172 покупки,а звичайні клієнти частіше використовують електронний гаманець 184 покупки. При цьому різниця між способами оплати не дуже велика*.
*Цю інформацію можна використати для підготовки акцій або бонусів під різні групи клієнтів. Наприклад, для учасників програми лояльності можна робити додаткові пропозиції при оплаті кредитною карткою, а для звичайних клієнтів — при оплаті електронним гаманцем*.

2.   **Які категорії товарів дають найбільший обсяг продажів і валовий дохід?**

*Найбільшу суму продажів має категорія Food and beverages 56 144,84 вона ж має найбільший валовий дохід 2 673,56. При цьому найбільша кількість товарів була продана в категорії Electronic accessories 971 одиниць,найнижчі продажі та валовий дохід має категорія Health and beauty*.
*Ці результати можна використати при плануванні запасів товару. Для категорій, які мають високі продажі варто підтримувати достатню кількість товару на складі. Також можна окремо проаналізувати категорії з нижчими продажами та вирішити чи потрібно змінювати їх асортимент або запланувати додаткові акції*.


3.  **Чи змінюється задоволеність клієнта залежно від розміру покупки?**

*Середній рейтинг майже однаковий для всіх розмірів покупки: Small 6,97; Medium 7,01; Large  6,93. Тому за отриманими даними не видно значної залежності між кількістю товарів у покупці та задоволеністю клієнта. При цьому середня сума покупки закономірно збільшується: від 114,40 для Small до 537,26 для Large*.
*Для бізнесу це означає, що великі покупки самі по собі не призводять до помітного зниження рейтингу. Тому немає явної потреби окремо змінювати обслуговування клієнтів з великими покупками, але їх можна додатково стимулювати бонусами або програмою лояльності так як вони залишають у магазині більшу суму*.






